New Rule vue/sort-keys

loren138 opened the pull request:
This rule creates a version of sort-keys which is compatible with `order-in-components`.

This addresses the issue raised in #286 as well fulfilling the request in #601.

loren138 commented on .eslintignore:
```diff
@@ -2,7 +2,7 @@
 /node_modules
 /tests/fixtures
-/tests/integrations/*/node_modules
+/tests/integrations/eslint-plugin-import
```
EsLint 6.7.0 changes eslintignore parsing.  The previous one seems to have ignored the whole folder which I'm now explicitly doing.  Otherwise, you get errors about eslint-plugin-ignore not being installed when running `npm run lint` after a clean install.

ota-meshi commented:
Thank you for this PR.

Could you change the to ignore the following Vue component options?

- props.xxx  
  e.g.

  ```js
  props: {
    foo: {
      type: String,
      default: 'foo'
    }
  }
  ```

- computed.xxx  
  e.g.

  ```js
  computed: {
    foo: {
      get() {},
      set(foo) {}
    }
  }
  ```

- watch.xxx  
  e.g.

  ```js
  watch: {
    foo: {
      handler() {},
      deep: true,
    }
  }
  ```

- directives.xxx  
  e.g.

  ```js
  directives: {
    focus: {
      bind() {},
      inserted() {}
    }
  }
  ```

- components.xxx
- mixins[*].xxx
- extends
- inject.xxx

- model  
  e.g.

  ```js
  model: {
    prop: 'checked',
    event: 'change'
  }
  ```

loren138 commented:
Hi,
I'm not sure I completely understand your request.

My goal was to make sure all the props, computed, data, methods, etc are in alphabetical order while allowing the component order to be set by `vue/order-in-components`.

That would make this example invalid
```
props: {
  foo: {
    type: String,
    default: 'foo'
  }
}
```

it would have to have default before type
```
props: {
  foo: {
    default: 'foo',
    type: String
  }
}
```

The main goal, however, was to make this invalid:
```
props: {
  foo: {
    default: 'foo',
    type: String
  },
  bar: {
    default: 'foo',
    type: String
  },
}
```

and require it to have `bar` before `foo`:
```
props: {
  bar: {
    default: 'foo',
    type: String
  },
  foo: {
    default: 'foo',
    type: String
  }
}
```

ota-meshi commented:
Hi! @loren138 .

It doesn't exist now, but we may add rules to enforce the order of props definition in the future. This already has an issue open #541.

I think you need to ignore the some options in the Vue component so that they don't conflict when adding new rules.

loren138 commented:
In that case, what things would be enforced by this rule?  Can you give an example of something that you think would be invalid (while also ignoring all the things you mentioned above)?

ota-meshi commented:
I think rule #541 will report an error with the following code:

```js
props: {
  foo: {
    default: 'foo',
    type: String // The `type` property should come before the `default` property.
  },
}
```

The following code is valid:

```js
props: {
  foo: {
    type: String,
    default: 'foo'
  },
}
```

The following code does not report an error with rule #541.
However, the `vue/sort-keys` rule reports an error.

```js
props: {
  foo: { },
  bar: { },
}
```

loren138 commented:
Got it, I'll work on that.

loren138 commented:
@ota-meshi I think the latest commit addresses your comments.  I made what to ignore configurable in case people don't want to use any of the new rules or want to force alphabetical order until those are ready.

ota-meshi commented on lib/rules/sort-keys.js:
```diff
@@ -0,0 +1,239 @@
+    const reportErrors = (isVue) => {
+      if (isVue) {
+        errors = errors.filter((error) => {
+          let filter = error.hasUpper
```
Vue component objects may not be top-level objects. Excluding it with hasUpper can cause false positives.

```js
const obj = {
  foo() {
    Vue.component('my-component', {
      name: 'app',
      data() {}
    })
  }
}
```

ota-meshi commented on lib/rules/sort-keys.js:
```diff
@@ -0,0 +1,239 @@
+            message: "Expected object keys to be in {{natual}}{{insensitive}}{{order}}ending order. '{{thisName}}' should be before '{{prevName}}'.",
+            parentName: stack.upper && stack.upper.prevName,
+            grandparentName: stack.upper && stack.upper.upper && stack.upper.upper.prevName,
+            greatGrandparentName: stack.upper && stack.upper.upper && stack.upper.upper.upper && stack.upper.upper.upper.prevName,
```
It may false negatives if the objects are not chained.

```js
export default {
  computed: {
    foo () {
      return {
        b,
        a
      }
    }
  }
}
```

ota-meshi commented on docs/rules/sort-keys.md:
````diff
@@ -0,0 +1,103 @@
+
+```json
+{
+    "sort-keys": ["error", "asc", {"caseSensitive": true, "natural": false, "minKeys": 2}]
````
Could you include the properties added to the example?

loren138 commented:
@ota-meshi Thanks for catching those issues.  They are fixed now.  Let me know if you see anymore.

loren138 commented:
@ota-meshi Any further comments on this PR?

ota-meshi commented:
@loren138  sorry, but I haven't checked it yet.
I'm a little busy now, so please wait.
